etest_test/build_test.py

```diff
@@ -9,4 +14,193 @@ def test_help() -> None:
+    )
+    run = mocker.patch(
+        "etest.build.docker.container.run",
+        return_value=True,
+    )
+    commit = mocker.patch("etest.build.docker.container.commit")
+
+    # Mock cleanup functions
+    image_remove = mocker.patch("etest.build.docker.image.remove")
+    container_remove = mocker.patch("etest.build.docker.container.remove")
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--arch", arch, "--verbosity", "DEBUG"])
+
+    # Check for errors
+    assert result.exit_code == 0
+
+    # Assert all functions have been called
+    build.assert_called_once()
+    run.assert_called_once()
+    commit.assert_called_once()
+    image_remove.assert_called_once()
+    container_remove.assert_called_once()
+
+
+@pytest.mark.parametrize("libc", ["glibc", "musl", "uclibc"])
+@pytest.mark.parametrize("hardened", ["--hardened", "--no-hardened"])
+def test_alternate_libc_build(libc: str, hardened: str, mocker: MockerFixture) -> None:
+    """Ensure images with alternate libcs build correctly."""
+    # Mock build functions
+    build = mocker.patch(
+        "etest.build.docker.image.build",
+        return_value=True,
+    )
+    run = mocker.patch(
+        "etest.build.docker.container.run",
+        return_value=True,
+    )
+    commit = mocker.patch("etest.build.docker.container.commit")
+
+    # Mock cleanup functions
+    image_remove = mocker.patch("etest.build.docker.image.remove")
+    container_remove = mocker.patch("etest.build.docker.container.remove")
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--libc", libc, hardened, "--verbosity", "DEBUG"])
+
+    # Check for errors
     assert result.exit_code == 0
+
+    # Assert all functions have been called
+    build.assert_called_once()
+    run.assert_called_once()
+    commit.assert_called_once()
+    image_remove.assert_called_once()
+    container_remove.assert_called_once()
+
+
+def test_builderror(mocker: MockerFixture) -> None:
+    """Ensure the program fails on a BuildError."""
+    # Raise a ContainerError
+    log = itertools.tee({"error": "mockup"})[0]
+    error = docker.errors.BuildError(reason="Mockup error.", build_log=log)
+
+    build = mocker.patch(
+        "etest.build.docker.image.build",
+        side_effect=error,
+    )
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--verbosity", "DEBUG"])
+
+    # Check if it triggered an error
+    assert result.exit_code == 1
+
+    # Assert all functions have been called
+    build.assert_called_once()
+
+
+def test_cleanup(mocker: MockerFixture) -> None:
+    """Ensure cleaning happens after exceptions."""
+    # Mock build functions
+    build = mocker.patch(
+        "etest.build.docker.image.build",
+        return_value=True,
+    )
+
+    # Raise a ContainerError
+    error = docker.errors.ContainerError(
+        container=True,
```

Comment:
As we dont actually do anything with the error container when all of the mocks are in place, we just need them to be truthy in case it tests if it exists at some point. I agree it is a weird type though. What could be a better option in your opinion?

Reply:
What type does container have?  Is True a legal value for that type?  If not, then we should make the code type correct and go from there.  I always try to use zero values for stubs and then inject the values I need onto the zero.  container=True reads as a type error to me which is why I ask.

Reply:
Intuitively it should have a type of ``docker.models.containers.Container`` but in reality it doesnt have any type hints (https://github.com/docker/docker-py/blob/96c12726fdb9b24e985b4de74d5c82145ccd8185/docker/errors.py#L125). Im unsure as to how to create this type correctly without putting extra computing power by actually creating a container

Reply:
If you're just worried about creating a stub, a MagicMock would work just fine for now and is understood by the type system.  That'll be much more clear if you use docker.models.containers.Container as the spec for the mock as well.